src/openslide-vendor-dicom.c

```diff
@@ -0,0 +1,932 @@
+    print_frame(frame);
+
+    if (!_openslide_jpeg_decode_buffer(frame_value,
+                                       frame_length,
+                                       box.p,
+                                       dcm_frame_get_columns(frame),
+                                       dcm_frame_get_rows(frame),
+                                       err)) {
+      return false;
+    }
+
+    // clip, if necessary
+    if (!_openslide_clip_tile(box.p,
+                              tile_size, tile_size,
+                              l->base.w - tile_col * tile_size,
+                              l->base.h - tile_row * tile_size,
+                              err)) {
+      return false;
+    }
+
+    // put it in the cache
+    tiledata = _openslide_slice_steal(&box);
+    _openslide_cache_put(osr->cache,
+			 level, tile_col, tile_row,
+			 tiledata, tile_size * tile_size * 4,
+			 &cache_entry);
+  }
+
+  // draw it
+  g_autoptr(cairo_surface_t) surface =
+    cairo_image_surface_create_for_data((unsigned char *) tiledata,
+                                        CAIRO_FORMAT_ARGB32,
+                                        tile_size, tile_size, tile_size * 4);
+  cairo_set_source_surface(cr, surface, 0, 0);
+  cairo_paint(cr);
+
+  return true;
+}
+
+static bool paint_region(openslide_t *osr, 
+                         cairo_t *cr,
+                         int64_t x, 
+                         int64_t y,
+                         struct _openslide_level *level,
+                         int32_t w, 
+                         int32_t h,
+                         GError **err) {
+  struct dicom_ops_data *data = osr->data;
+  struct dicom_level *l = (struct dicom_level *) level;
+
+  g_debug("paint_region: x = %" PRId64 ", y = %" PRId64 ", w = %d, h = %d", 
+          x, y, w, h);
+  g_debug("paint_region level:");
+  print_level(l);
+
+  return _openslide_grid_paint_region(l->grid, 
+                                      cr, 
+                                      data,
+                                      x / l->base.downsample,
+                                      y / l->base.downsample,
+                                      level, 
+                                      w, 
+                                      h,
+                                      err);
+}
+
+static const struct _openslide_ops dicom_ops = {
+  .paint_region = paint_region,
+  .destroy = destroy,
+};
+
+static bool is_type(struct dicom_file *f, const struct allowed_types *types) {
+  // ImageType must be one of the combinations we accept
+  int match = -1;
+  for (int i = 0; i < types->n_types; i++) {
+    bool found_difference = false;
+
+    for (int j = 0; j < 4; j++) {
+      const char *type;
+      if (!get_tag_str(f->metadata, "ImageType", j, &type)) {
+        return false;
+      }
+
+      if (!g_str_equal(type, types->types[i][j])) {
+        found_difference = true;
+        break;
+      }
+    }
+
+    if (!found_difference) {
+      match = i;
```

Comment:
We could just `return true` here and drop `match`.

Reply:
OK, changed.